server/stats.go

```diff
@@ -6,18 +6,56 @@
 package main
 
 import (
+	"encoding/json"
 	"expvar"
 	"log"
 	"net/http"
 	"runtime"
 	"time"
 )
 
+// A simple implementation of histogram expvar.Var.
+// `Bounds` specifies the histogram buckets as follows (length = len(bounds)):
+//     (-inf, Bounds[i]) for i = 0
+//     [Bounds[i-1], Bounds[i]) for 0 < i < length
+//     [Bounds[i-1], +inf) for i = length
+type histogram struct {
+	Count          int64     `json:"count"`
+	CountPerBucket []int64   `json:"count_per_bucket"`
+	Bounds         []float64 `json:"bounds"`
+}
+
+func (h *histogram) addSample(v float64) {
+	h.Count++
+	var count *int64
+	var i int
+	var b float64
+	for i, b = range h.Bounds {
+		if v < b {
+			count = &h.CountPerBucket[i]
+			break
+		}
+	}
+	if count == nil { // Last bucket.
+		i = len(h.Bounds)
+		count = &h.CountPerBucket[i]
+	}
+	*count++
+}
+
+func (h *histogram) String() string {
+	if r, err := json.Marshal(*h); err == nil {
```

Comment:
Do you need to dereference the struct and pass it by value? It will create a copy of a potentially large struct.

Reply:
Done.